### DISCUSSION TITLE
Snapmaker J1 Q&A

### DISCUSSION THREAD

**Discussion by hliebscher** [OrcaSlicer/OrcaSlicer#782] [General]
this is a place for Q&A for my Snapmaker integration

@theplanekid @Learoy1 we can make a discussion  
i open this
mirror and copy mode can activated in gcode
As long as the model you have sliced is less than half the size of the bed you can mirror and copy on the printer. 
on print bed left side behind the blue or red line
/Users/hliebscher/smfix_bs-darwin-arm64;  modify for Orca
/Users/hliebscher/j1x2;  send the file to J1. 
i use my own software to send the gcode to the j1 and fix the code for display the values

**Comment by Learoy1** [comment]
Perfect, can't wait to try it out. 
I recently got a notification that sm2uploader has been updated to simulate octoprint for sending via WiFi and not requiring to authorise on the printer. Not sure if it's needed or helps but linky...
https://github.com/macdylan/sm2uploader

**Reply by theplanekid** [reply]
do you know how to use sm2uploader to simulate octoprint? i run the executable in terminal and it doesn't discover any printers, but luban is able to detect it.
please advise, thank you!

**Reply by hliebscher** [reply]
smuploader works fine but not for 4 printers

**Reply by hliebscher** [reply]
[Releases 1](https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor/releases)

**Comment by theplanekid** [comment]
smfix_bs-darwin is from from here correct? https://forum.snapmaker.com/t/great-sharing-prusaslicer-profiles-by-dylan-and-mcgybeer/28796
would it be possible to get the j1x2 software?

**Reply by hliebscher** [reply]
> smfix_bs-darwin is from from here correct? https://forum.snapmaker.com/t/great-sharing-prusaslicer-profiles-by-dylan-and-mcgybeer/28796 

Yes but this version works only for Prusa my works only for OrcaSlicer

would it be possible to get the j1x2 software?

if you have only 1 Snapmaker in Network it is easy 
if you have more than you need the ipaddresswitch 
witch version do yo need mac / win ?

**Reply by theplanekid** [reply]
oh, where can we find your version of smfix then?
i don’t have multiple snapmakers, so i guess i don’t need the j1x2 software.

**Reply by hliebscher** [reply]
https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor.git
and select OrcaSlicer Branch

witch version do yo need?
@projectsaturnstudios @theplanekid @Learoy1 @arciarc

**Comment by szaba1de** [comment]
Where can i chose witch extruder want i use?

**Reply by hliebscher** [reply]
you have to add a second color than switch the extruder automatic

**Comment by szaba1de** [comment]
![image](https://user-images.githubusercontent.com/120054892/233630490-adcef4a9-6c66-4eb9-bf61-2e40e332b234.png)
![image](https://user-images.githubusercontent.com/120054892/233630540-4a8479b7-5a6f-407f-9e1b-eae7c4970b11.png)
Where can I find the files and why is the path specified that way?

**Reply by hliebscher** [reply]
witch OS ?

**Reply by hliebscher** [reply]
https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor.git
and select OrcaSlicer Branch

**Reply by hliebscher** [reply]
[Releases 1](https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor/releases)

**Reply by szaba1de** [reply]
Thanks its work!

**Comment by hliebscher** [comment]
https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor.git
and select OrcaSlicer Branch

witch version do yo need?

see at discussion Board in OrcaSlicer

@projectsaturnstudios @theplanekid @Learoy1 @arciarc

**Reply by hliebscher** [reply]
[Releases 1](https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor/releases)

**Comment by Frelidon** [comment]
Unfortunately it does not work.
Would someone be so kind as to explain it to me on discord.

**Reply by hliebscher** [reply]
@Frelidon We can call in German next week

**Reply by Frelidon** [reply]
OK Danke gerne, würde ihn gerne verwenden.

**Reply by arciarc** [reply]
Hallo, da würde ich mich gerne mit anhängen.
Vielleicht schafft es jemand (oder ich, wenn ich es denn zum Laufen bringe), eine "Anleitung für EDV-Dummies" zu schreiben, damit das klappt. Wäre schade um die Arbeit vom Orca-Slicer-Team und Heiko Liebscher @hliebscher, wenn das keine größere Verbreitung finden könnte.

**Comment by Learoy1** [comment]
OK, I finally get chance to try. 
Everything is looking great on Orca slicer but smfix is causing issues and cannot generate gcode. 
I downloaded your smfix and place it in C:\Program Files\smfix\smfix_bs-win64.exe , I copy paste that exact path and place in the print settings post processing script. once i export gcode I then get error "the post processing script does not exist" 

Also, where is j1x2? where do I download that? 

I appreciate your work and I know this is my user error but there are no clear instructions for an inexperienced person to set this up.

I have a Snapmaker J1 and a Snapmaker 2 so need to send files to 2 different snapmaker printers, is this an issue?

**Reply by hliebscher** [reply]
i write a short description in german and english ... i have 2 J1 and 2 SM2 so i have my own j1x2 with the ip from the first printer .... and so on  t hink tomorrow it is possible ... Thanks

**Comment by fr3ak2402** [comment]
Hi @hliebscher, I have a few questions about the Snapmaker J1 in OrcaSlicer.

I download the post processor from the following link?
https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor/releases

After that I have to adjust my G-code so that the Post Processor points to my paths?

In the start code I noticed that you use parameters there that I don't find in OrcaSlicer:
"first_layer_bed_temperature", "first_layer_temperature".

I can find the following values in OrcaSlicer, "hot_plate_temp_initial_layer" and "nozzle_temperature_initial_layer".

Consequence from these values is that the bed heats despite Temp setting of 60°C only to 45°C and with each ToolChange the temperature of the nozzle, which one gives at the filament is lowered again to 200°C. But I don't know where it gets the 45°C and the 200°C from.

Do I also have to adjust these parameters so that the temperatures are taken over correctly during the printing process?

first_layer_bed_temperature -> hot_plate_temp_initial_layer
first_layer_temperature -> nozzle_temperature_initial_layer

Furthermore, for ToolChanges >0, the second hotend does not heat from the beginning, but only during the ToolChange itself and does not wait for the temperature, but starts immediately while still cold. Is this also due to the start code?

If I have understood this correctly, to be able to print with two colors. Do I just have to set another filament in OrcaSlicer and the printer will automatically assign them to the two extenders?

You are also welcome to answer me in german if you like....

**Reply by hliebscher** [reply]
sorry for all others write in  German later for all in eng

@fr3ak2402 
das funktioniert eigentlich perfekt... alle variablen die ich im start script habe existieren 
   // SoftFever: write compatiple image. haben wir extra eingebaut
        int first_layer_bed_temperature = get_bed_temperature(0, true, print.config().curr_bed_type);
        file.write_format("; first_layer_bed_temperature = %d\n",
                          first_layer_bed_temperature);
        file.write_format(
            "; first_layer_temperature = %d\n",
            print.config().nozzle_temperature_initial_layer.get_at(0));
        file.write("; CONFIG_BLOCK_END\n\n");

wichtig ist der Start Code und der End Code im Printer
zum testen auf die variablen einfach mit dem richtigen Platzhalter in den start code zum testen

aber das wichtigste ist der G Code im Filament ! 
es müssen immer 2 Filamente im Orca angegeben sein und die müssen beide meinen G Code enthalten!
unterschiedliche Temperaturen gehen
wichtig ist der Prime Tower muss aktiv sein!
Wenn du eigene Filament Einstellungen verwendest muß immer cer Start Code und End Code von mir mit rein !
Ja der Pfad muß im Postprocess angepaßt werden
die 2. Datei schickt den geänderten G-Code direkt an den Drucker.
Da kannst du sm2send oder andere verwenden.
bei mir sind die IP's der Drucker fest eingestellt da ein discover nicht geht (Timeout) 
habe 4 Drucker und da dauert die Antwort zu lang...


;***** Snapmaker J1
;***** Update  : 20221222
;***** Printer : [printer_preset]
;***** Profile : [print_preset]
M593  P1 X F35.7  
M593  P1 Y F33.7 
M355 S1 P16    ;------------OPTIONAL                ;dim led 
M83 ;absolute extrusion mode
M140 S[first_layer_bed_temperature] ;Start Warming Bed
T[initial_extruder] ;switch to initial extruder <------Hier wird der rechte oder linke  aufgeheizt
{if total_toolchanges < 1}; Preheat one extruder (just one used)
M104 S160 ;preheat nozzle. auf 160 Grad vorheizen während Bed aufheizt
{else}; Preheat both extruders used
M104 S160 T0 ;preheat T0 beide auf 160 Grad vorheizen während Bed aufheizt
M104 S160 T1 ;preheat T1
{endif}
G28 ; home all axes
G90 ;absolute positioning
G1 Z0.8 F1800
M201 X5000 Y5000 Z500 E5000
M205 V5
M900 T0 K0.02
M900 T1 K0.02
M190 S[first_layer_bed_temperature]  ; C3 W2 Wait For Bed Temperature
{if total_toolchanges < 1}; Heat one extruder
M104 S[first_layer_temperature] 
M109 S[first_layer_temperature] ;Wait for Hotend Temperature
{else}; Heat both extruders
M104 T0 S{first_layer_temperature[0]}
M104 T1 S{first_layer_temperature[1]}
M109 T0 S{first_layer_temperature[0]} ;Wait for Hotend Temperature
M109 T1 S{first_layer_temperature[1]} ;Wait for Hotend Temperature
{endif}
;
{if 1==1}      ;-----------OPTIONAL                 ;led effect
  M355 S1 P32
  G4 S1
  M355 S1 P48
  G4 S1
  M355 S1 P64
  G4 S1
  M355 S1 P96
  G4 S1
  M355 S1 P128
  G4 S1
  M355 S1 P255
{endif}
;
G92 E0
;
{if total_toolchanges > 0 or initial_extruder == 0}; Purge line T0
T0
G92 E0
G1 F1200 X5 Z0.5
G1 F600 X100 E20
G1 F600 Z2
G92 E0
G1 F1200 E-6
G0 X-20 F12000 ; park T0
{endif}
{if total_toolchanges > 0 or initial_extruder == 1}; Purge line T1
T1
G92 E0
G1 F1200 X295 Z0.5
G1 F600 X200 E20
G1 F600 Z2
G92 E0
G1 F1200 E-6
G0 X350 F12000 ; park T1 (old 345)
{endif}
;End Start Gcode J1


und hier der G-Code  nach dem slicen

;TYPE:Custom
;***** Snapmaker J1
;***** Update  : 20221222
;***** Printer : Snapmaker J1 0.6 Düse
;***** Profile : 0.20mm 0.6 Standard @Snapmaker J1 - MK
M593  P1 X F35.7 
M593  P1 Y F33.7 
M355 S1 P16    ;------------OPTIONAL                ;dim led 
M83 ;absolute extrusion mode
M140 S60 ;Start Warming Bed
T0 ;switch to initial extruder
; Preheat both extruders used
M104 S160 T0 ;preheat T0
M104 S160 T1 ;preheat T1

G28 ; home all axes
G90 ;absolute positioning
G1 Z0.8 F1800
M201 X5000 Y5000 Z500 E5000
M205 V5
M900 T0 K0.02
M900 T1 K0.02
M190 S60  ; C3 W2 Wait For Bed Temperature
; Heat both extruders
M104 T0 S215
M104 T1 S215
M109 T0 S215 ;Wait for Hotend Temperature
M109 T1 S215 ;Wait for Hotend Temperature

;
      ;-----------OPTIONAL                 ;led effect
  M355 S1 P32
  G4 S1
  M355 S1 P48
  G4 S1
  M355 S1 P64
  G4 S1
  M355 S1 P96
  G4 S1
  M355 S1 P128
  G4 S1
  M355 S1 P255

;
G92 E0
;
; Purge line T0
T0
G92 E0
G1 F1200 X5 Z0.5
G1 F600 X100 E20
G1 F600 Z2
G92 E0
G1 F1200 E-6
G0 X-20 F12000 ; park T0

; Purge line T1
T1
G92 E0
G1 F1200 X295 Z0.5
G1 F600 X200 E20
G1 F600 Z2
G92 E0
G1 F1200 E-6
G0 X350 F12000 ; park T1 (old 345)

;End Start Gcode J1

**Reply by arciarc** [reply]
Hallo,

ich versuche gerade wieder einmal das "generate G-Code" zum Laufen zu bringen.
Leider bekmme weiterhin die Fehlermeldung, dass das Script nich existiert.
"/Benutzer/mein-Benutzer-Name-Ordner/AppData/Roaming/OrcaSlicer/user/smfix_bs-darwin-arm64"

Ich habe als Filament1 das SnapmakerJ1 PLA und als Filament2 das SnapmakerJ1 PVA gewählt (für Support)
Bin auf Win10 x64. 

OrcaSlicer mit Bambu P1P & X1C geht bestens.

Muss ich da nun noch bei den Filamenten G-Code ändern? 
Es gibt ja jeweils den Start-Code und den End-Code?

Wahrscheinlich übersehe ich eine Kleinigkeit, aber ich bin nicht so oft mit Code bzw. Programmierung befasst, sondern ändere meist nur einzelne Werte, wenn ich das Grundgerüst jeweils verstanden habe.

Ich würde mich sehr freuen, wenn ich das schaffen würde. Das direkte "Senden an Drucker" ist dann der nächste Schritt, aber ich kann ja auch einen USB-Stick rübertragen.

Danke Euch für Tipps, wo der Fehler liegt.
arciarc (Boris)

**Reply by fr3ak2402** [reply]
@arciarc 

Hast du einen Windows 10 Arm oder ein Windows x64?

Wenn du ein Windows x64 hast, musst du eine andere Datei herunterladen.

https://github.com/hliebscher/Snapmaker2Slic3rPostProcessor/releases

<img width="957" alt="Bildschirmfoto 2023-05-09 um 21 48 19" src="https://github.com/SoftFever/OrcaSlicer/assets/96239814/895384c9-7c3c-4566-874d-562d819ff7d9">

Denn dann passt sie auch zum Betriebssystem.

Dann kopiere die Datei in den folgenden Ordner "/Benutzer/Mein-Benutzername-Ordner/AppData/Roaming/OrcaSlicer/user/".

Danach muss nur noch der Pfad im OrcaSlicer angepasst werden. Gehe auf den Reiter " Sonstiges" und scrolle ganz nach unten, bis ein Eingabefeld für die Post-Processing-Skripte erscheint. In diesem Feld ersetze einfach den Pfad durch deinen Pfad.

Dann sollte das Slicen bei dir funktionieren.

**Reply by arciarc** [reply]
Danke, so habe ich das gemacht. Slicen geht, aber ich kann die Datei nicht Exportieren oder Drucken.
Egal ob ich im Post-Sctript-Feld das Script mit *.exe oder ohne Dateiendung angebe, es kommt immer die selbe Meldung "Script nicht gefunden"

Hier das gesamte im Feld eingegebene Script:
C:/Benutzer/Mein-Benutzername-Ordner/AppData/Roaming/OrcaSlicer/user/smfix_bs-win64.exe;
getestet mit und ohne Dateieendung
getestet mit und ohne das C: vornedran

Windows benutzt ja traditionell \ und nicht /  aber das ist es auch nicht, oder?

Ich stelle mich normalerweise bei derartigen Anpassungen nicht so doof an, aber diesmal schaffe ich es nicht. Die Meldung kommt immer beim Export des G-Code.
"The configured post processing script does not exist - s.o."

Ich hoffe, das ich es zum Laufen bekomme. 
Dann auch mit dem Drucken über Wlan ;-)

**Reply by fr3ak2402** [reply]
@arciarc 

Ja, dies wird auf einem Unix/Linux & macOS System verwendet. Du solltest dies auf einen Windows OS Pfad anpassen. Das wird auch der Fehler sein, dass er das Skript nicht findet.

Wenn du möchtest, können wir uns das gemeinsam auf Discord anschauen.

**Reply by arciarc** [reply]
Sehr gerne. Ich habe Discord noch nie benutzt, aber wird schon klappen.

Das Ändern von "/" zu "\" hat leider auch nichts genutzt.

**Reply by arciarc** [reply]
auf Discord selber Benutzername.
LG

**Reply by fr3ak2402** [reply]
Gibst du mir auch die #Nummer deines Discord-Namens? Sonst kann ich dich nicht finden.

**Reply by arciarc** [reply]
jetzt, hier: 4527

**Reply by arciarc** [reply]
sorry, ich bin dauernd unterwegs im Atelier, jetzt aber mal kurz da. 
Sonst einfach ein anderes Mal, wenn´s besser passt. 
Sobald Du mich anpingst müsste das auf Discord aufpoppen, oder?

**Reply by hliebscher** [reply]
@arciarc kannst du die datei im Datei-Explorer ausführen?
wenn ja den Pfad oben direkt aus dem Fenster kopieren .... vielleicht hast du ein Leerzeichen im Namen?
dann muss der Pfad angepasst werden oder in "" geschrieben werden
und zum senden bitte noch die 
https://github.com/macdylan/sm2uploader einbinden

**Reply by arciarc** [reply]
Hallo hliebscher,
den G-Code kann ich nun exportieren. fr3ak2402 hat mir geholfen und auf die Idee, dass leider in einem Ordnernamen Leerzeichen sind wir dann auch gekommen.
Den sm2uploader versuche ich jetzt auch.
...und es hat geklappt! 
Am Drucker Mirrormode auswählen und zwei Kreisel drucken jetzt parallel :-)
Bin gespannt, wie ich Dual-Material-Prints schaffe.

Nochmal Danke für Eure Hilfe und Eure Programmierkunst!

**Comment by arciarc** [comment]
For the international community:
You will find the place to put the post-processing-scripts on the left side:
-> others, scroll down to bottom, insert your path including script-name(s)
-> save the changes not with the printersettings, but with the settings for your printprofile (yes, if you use other presets, you´ll have to safe the script again)
If you put the path to your scripts take care that there is no space inm any of the filenames or the path/ folders. 
(or, as suggested by hliebscher: put the path in " " - not tested by me)
For Windows use backslash '\' and not '/' for path-separation. 
Also add the hard disk letter at the beginning, and fileextension (*.exe):
C.\yourpath\yourscript.exe
last:
don´t forget to use Snapmaker-Filament-Profiles only (feel free to adjust for your needs) and always take two of them, as the first will be left extruder and the second will be right extruder.

I hope this short instructions help keeping the question-rate low.
:-)
Enjoy your Snapmaker J1

**Reply by Belvis87** [reply]
![Screenshot 2023-11-24 082640](https://github.com/SoftFever/OrcaSlicer/assets/151910136/c949cc22-5dfa-46b7-a853-196f6da17689)
what am I doing wrong?

**Reply by Learoy1** [reply]
> ![Screenshot 2023-11-24 082640](https://github.com/SoftFever/OrcaSlicer/assets/151910136/c949cc22-5dfa-46b7-a853-196f6da17689)
> what am I doing wrong?

I'm not an expert but there is a mixture of backslash and forward slash in your address, that might be a problem

**Reply by Learoy1** [reply]
> what am I doing wrong?

It also looked like you have no .exe file extension at the end of your address. 
Re read and follow the main instructions above.

**Comment by hliebscher** [comment]
https://github.com/Snapmaker/SnapmakerController-IDEX

**Reply by Learoy1** [reply]
It's great that they have now released this. Will it be useful for any of your work?

**Comment by Bauerranger10** [comment]
@hliebscher
I've got everything working so far and the prints look very good too. Unfortunately, printing with 2 colors is difficult because apparently Orca is not designed for 2 extruders. I noticed when you print 2 different materials that the temperatures get swapped. Am I doing something wrong or do we have to wait for an update?
Thanks in advance

@hliebscher 
Hab so weit alles zum Laufen gebracht und die Drucke sehen auch sehr gut aus. Leider stellt sich das Drucken mit 2 Farben als schwierig dar, da anscheinend Orca nicht ausgelegt ist für 2 Extruder. Mir ist aufgefallen, wenn man 2 unterschiedliche Materialien druckt, dass die Temperaturen vertauscht werden. Mach ich was falsch oder müssen wir da auf ein Update warten?
Danke schonmal

**Comment by TylonHH** [comment]
Can someone please tell me where are the build plates for choosing to print in mirror or copy mode?
![grafik](https://github.com/user-attachments/assets/cb03c45e-fd42-49e8-a79e-7c9312f5c821)

**Reply by Learoy1** [reply]
You have to activate mirror or copy mode on the printers touchscreen when you select the print from the USB drive or internal storage. It's a while since I used the J1 POS but once you've selected WHAT you want to print, there will be a button somewhere to select between print modes. 
In the slicer just make sure your model is smller than half the build plate, and position it on the left. I think the slicer should have the lines on the build plate to show you the limits.

**Reply by TylonHH** [reply]
Ahh ok. Got it. That makes sense.
Unfortunately I print via octoprint, so I have manually set the mirror mode, but that's no problem.
Thanks for the quick response.
